### PR TITLE
FiNER Wallet support

### PR DESCRIPTION
**Pull request by prih**
# Description

Add support of [FiNER Wallet](https://finerwallet.io) which can be installed from [App Store](https://apps.apple.com/us/app/finer-wallet/id1623425017) and [Chrome Store](https://chrome.google.com/webstore/detail/finer-wallet/mbboloafhcopdfokibemggdnhcocggpl)

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings

### CODE REVIEW

**Comment by prih** [comment]
@AmmarHumackicSQA @amirsaran3 We reviewed all the nuances of interaction with the wallet on the mobile device, also added integration with the browser expansion

**Comment by AmmarHumackicSQA** [comment]
> @AmmarHumackicSQA @amirsaran3 We reviewed all the nuances of interaction with the wallet on the mobile device, also added integration with the browser expansion

Thanks @prih, we will be checking it today.

**Comment by DamirSQA** [comment]
Hi @prih , thanks for submitting this PR. 
I've noticed some issues that you might want to check out. 

On desktop view, the Finer wallet isn't visible in the list in case it's not installed. It should be present with the install page when selected. You can have a look at the example below. 

image

In mobile view it is visible, but it throws this error when clicked.
image

After installing extension, when I'm trying to use wallet I'm getting "user reject" error when clicked on next. 
You can check out [video example](https://www.loom.com/share/110ab09be45c44b98505d6bdf7d103b4)

**Comment by prih** [comment]
> On desktop view, the Finer wallet isn't visible in the list in case it's not installed. It should be present with the install page when selected. You can have a look at the example below.

- fixed

**Comment by prih** [comment]
> After installing extension, when I'm trying to use wallet I'm getting "user reject" error when clicked on next.

its in process, we are waiting for the renewal of expansion in the store... v0.1.4 - will open the extension window and inform the user about the need to connect the expansion with the mobile application.

**Comment by prih** [comment]
> In mobile view it is visible, but it throws this error when clicked.

this message is not at the JS code level, when trying to open a mobile application using the URL scheme finer://
simulation on the desktop browser (devtools) does not make it a real mobile browser)

**Comment by prih** [comment]
> > In mobile view it is visible, but it throws this error when clicked.
> 
> this message is not at the JS code level, when trying to open a mobile application using the URL scheme finer:// simulation on the desktop browser (devtools) does not make it a real mobile browser)

image

**Review by DamirSQA — CHANGES_REQUESTED** [review]

**Comment by kujtimprenkuSQA** [comment]
@prih this is my initial review. these are some of the issues:

- Your wallet (FiNER extension) is creating or overriding the existing `window.near` object which is conflicting with the `Sender` wallet.
- The code about communicating with the extension is confusing and `Sender` is mentioned in a  lot of places it's a bit misleading.
- `signAndSendTransaction` with a deposit is not working on the mobile wallet.

This PR is trying to combine 2 different wallet implementations for this reason we will need to decide if there is a need to split these into 2 different packages (wallets) like FiNER Wallet (mobile), FiNER Wallet(extension).

Could you have a look at the contributing docs too and make sure the FiNER wallet aligns with the criteria?
https://github.com/near/wallet-selector/blob/main/CONTRIBUTING.md#listing-criteria-for-third-party-wallet-on-wallet-selector

For the Extension can you see if it aligns with the NEP standard: https://github.com/near/NEPs/pull/408

**Comment by prih** [comment]
> * Your wallet (FiNER extension) is creating or overriding the existing `window.near` object which is conflicting with the `Sender` wallet.

no, extension doesn't overwrite window.near only if it's undefined
image

packages/finer-wallet uses window.finer.near directly...

**Comment by prih** [comment]
> > * Your wallet (FiNER extension) is creating or overriding the existing `window.near` object which is conflicting with the `Sender` wallet.
> 
> no, extension doesn't overwrite window.near only if it's undefined image
> 
> packages/finer-wallet uses window.finer.near directly...

image

**Comment by kujtimprenkuSQA** [comment]
> no, extension doesn't overwrite window.near only if it's undefined

One way to find out this strange behaviour is if you disable `Sender` in the extensions of your browser and try clicking the `Sender` option. Sender and Finer must be 2 separate things.

**Comment by prih** [comment]
> * `signAndSendTransaction` with a deposit is not working on the mobile wallet.

yes, we see the problem, it worked before)

File: packages/finer-wallet/package.json [hunk 1]
```diff
@@ -0,0 +1,4 @@
+{
+  "name": "@near-wallet-selector/finer-wallet",
+  "version": "7.5.0"
+}
```
[the remark this answers is not in the document]
Comment (prih) [reply packages/finer-wallet/package.json:28]: fixed

File: packages/finer-wallet/src/lib/finer-sender.ts [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+  Optional,
+} from "@near-wallet-selector/core";
+import { waitFor } from "@near-wallet-selector/core";
+import type { InjectedFinerSender } from "./injected-wallet";
+import icon from "./icon";
+
+declare global {
+  interface Window {
+    finer: {
+      near: InjectedFinerSender | undefined;
```
[the remark this answers is not in the document]
Comment (prih) [reply packages/finer-wallet/src/lib/finer-sender.ts]: fixed

File: packages/finer-wallet/src/lib/finer-sender.ts [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+    };
+  }
+}
+
+export interface SenderParams {
```
[the remark this answers is not in the document]
Comment (prih) [reply packages/finer-wallet/src/lib/finer-sender.ts]: fixed

File: packages/finer-wallet/src/lib/finer-sender.ts [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+  options,
+  metadata,
+  store,
+  provider,
+  emitter,
+  logger,
+}) => {
+  const _state = setupSenderState();
+
+  // const cleanup = () => {
```
[the remark this answers is not in the document]
Comment (prih) [reply packages/finer-wallet/src/lib/finer-sender.ts]: fixed

File: packages/finer-wallet/src/lib/finer-sender.ts [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+    },
+
+    signOut,
+
+    async getAccounts() {
+      return getAccounts();
+    },
+
+    async verifyOwner({ message }) {
+      logger.log("Sender:verifyOwner", { message });
```
[the remark this answers is not in the document]
Comment (prih) [reply packages/finer-wallet/src/lib/finer-sender.ts]: fixed

File: packages/finer-wallet/src/lib/finer-sender.ts [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+      metadata: {
+        name: "FiNER Wallet",
+        description: "Browser extension wallet built on NEAR.",
+        iconUrl,
+        downloadUrl: "",
+        // "https://chrome.google.com/webstore/detail/",
+        deprecated,
+        available: installed,
+      },
+      init: Sender,
```
[the remark this answers is not in the document]
Comment (prih) [reply packages/finer-wallet/src/lib/finer-sender.ts]: fixed

File: packages/finer-wallet/src/lib/finer-sender.ts [hunk 7]
```diff
@@ -0,0 +1,324 @@
+// based on @near-wallet-selector/sender
+
```
[the remark this answers is not in the document]
Comment (prih) [reply packages/finer-wallet/src/lib/finer-sender.ts]: fixed

File: packages/finer-wallet/src/lib/injected-wallet.ts [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+interface Transaction {
+  receiverId: string;
+  actions: Array<Action>;
+}
+
+interface RequestSignTransactionsParams {
+  transactions: Array<Transaction>;
+}
+
+interface SenderEvents {
```
[the remark this answers is not in the document]
Comment (prih) [reply packages/finer-wallet/src/lib/injected-wallet.ts]: fixed

File: packages/finer-wallet/src/lib/injected-wallet.ts [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+  signature: string;
+}
+
+interface SignMessageResponse {
+  method: "signMessage";
+  error?: string;
+  response?: SignMessageData;
+}
+
+export interface InjectedFinerSender {
```
[the remark this answers is not in the document]
Comment (prih) [reply packages/finer-wallet/src/lib/injected-wallet.ts]: fixed

**Comment by prih** [comment]
> > no, extension doesn't overwrite window.near only if it's undefined
> 
> One way to find out this strange behaviour is if you disable `Sender` in the extensions of your browser and try clicking the `Sender` option. Sender and Finer must be 2 separate things.

Yes, I understand, but what's the problem? are we prohibited from using this API? if the user install Sender Wallet, then it will work without problems...
this is implemented for projects that have not integrated wallet-selector and will not do it, but already have integration with browser-wallet and Sender (since they appeared before wallet-selector became mainstream), and are de facto standards, like for example in Ethereum world (and other EVM) MetaMask plays such a role with its window.ethereum provider, and no one is surprised by this fact that the user has the wallet that suits him, and not something with which he can connect to a specific dApp.

**Comment by prih** [comment]
> This PR is trying to combine 2 different wallet implementations for this reason we will need to decide if there is a need to split these into 2 different packages (wallets) like FiNER Wallet (mobile), FiNER Wallet(extension).

why should we do this? our extension is not a standalone wallet, but a gateway for interacting with a mobile application, user can use the desktop experience, while his keys are safe on the mobile device

**Comment by prih** [comment]
> For the Extension can you see if it aligns with the NEP standard: [near/NEPs#408](https://github.com/near/NEPs/pull/408)

yes, we are following this, when we make our implementation of the injected wallet for NEAR, we will definitely pay attention to these documents with recommendations, but at the moment I think this is not relevant, because. communication takes place via the wallet-selector interface

**Comment by kujtimprenkuSQA** [comment]
> > > no, extension doesn't overwrite window.near only if it's undefined
> > 
> > 
> > One way to find out this strange behaviour is if you disable `Sender` in the extensions of your browser and try clicking the `Sender` option. Sender and Finer must be 2 separate things.
> 
> Yes, I understand, but what's the problem? are we prohibited from using this API? if the user install Sender Wallet, then it will work without problems... this is implemented for projects that have not integrated wallet-selector and will not do it, but already have integration with browser-wallet and Sender (since they appeared before wallet-selector became mainstream), and are de facto standards, like for example in Ethereum world (and other EVM) MetaMask plays such a role with its window.ethereum provider, and no one is surprised by this fact that the user has the wallet that suits him, and not something with which he can connect to a specific dApp.

If sender is not installed and you click it in the wallet selector UI it should show the `WalletNotInstalled` view and not prompt the connection with FiNER.

Based on the specification here Sender will eventually have to move to `window.near.sender` if they want to align with the Standard: https://github.com/near/NEPs/blob/master/specs/Standards/Wallets/InjectedWallets.md#specification

**Comment by kujtimprenkuSQA** [comment]
> > This PR is trying to combine 2 different wallet implementations for this reason we will need to decide if there is a need to split these into 2 different packages (wallets) like FiNER Wallet (mobile), FiNER Wallet(extension).
> 
> why should we do this? our extension is not a standalone wallet, but a gateway for interacting with a mobile application, user can use the desktop experience, while his keys are safe on the mobile device

I see at first I thought the extension was a wallet too.

**Comment by prih** [comment]
> If sender is not installed and you click it in the wallet selector UI it should show the `WalletNotInstalled` view and not prompt the connection with FiNER.
> 
> Based on the specification here Sender will eventually have to move to `window.near.sender` if they want to align with the Standard: https://github.com/near/NEPs/blob/master/specs/Standards/Wallets/InjectedWallets.md#specification

**Comment by prih** [comment]
> If sender is not installed and you click it in the wallet selector UI it should show the `WalletNotInstalled` view and not prompt the connection with FiNER.

I agree, but this is done for compatibility, we can solve this by notifying the user that this is not a Sender and offering to use FiNER, if he refuses, then we will set a flag about this and will not offer FiNER, ok?

**Comment by prih** [comment]
> Based on the specification here Sender will eventually have to move to `window.near.sender` if they want to align with the Standard: https://github.com/near/NEPs/blob/master/specs/Standards/Wallets/InjectedWallets.md#specification

ok, it's up to them, we just want users to be able to replace window.near if the dApp can only do that...

**Comment by prih** [comment]
> * `signAndSendTransaction` with a deposit is not working on the mobile wallet.

it is interesting that it didn’t problem with us, the wallet is waiting for the finalization of the transaction, and it doesn’t happen in any way. node loses recipes... but in fact it is in the blockchain, just rpc for some reason does not associate it with this transaction...

example:
https://explorer.testnet.near.org/transactions/2z8WphkMiSXoz3tNrw6ehhWghbdSAVWB9JRwsJNSzv4p

we have made adjustments to our finalization check mechanism, it should work with the new app update...

**Comment by DamirSQA** [comment]
Hi @prih , thanks for the changes and updates. `signAndSendTransaction` is working now, one thing left to be addressed is the `windows.near` issue.

**Comment by prih** [comment]
> Hi @prih , thanks for the changes and updates. `signAndSendTransaction` is working now, one thing left to be addressed is the `windows.near` issue.

@DamirSQA @kujtimprenkuSQA 

what is the issue?) we do not see any problems and reasons to change anything, our extension is api compatible with Sender Wallet, and this is so for compatibility, there are dApps that have directly integrated this api, and do not use wallet-selector, or early versions of it.

ok, if you are not satisfied with how our extension works, then - what rules does he break?

**Comment by prih** [comment]
Is this the decentralization we deserve? why are we dictated how we should make a wallet and what api should it have? are we somehow conflicting with sender-wallet ? - no, only if it is not installed by the user, we are trying to make life easier for the user.

**Comment by prih** [comment]
> Hi @prih, the reason behind enforcing the standard is that we ensure everyone has the same conditions for their wallets without negative impact on each other.

what negative impact do we have? if the user has Sender installed, then our extension does not interfere with him in any way...

**Comment by AmmarHumackicSQA** [comment]
> > If sender is not installed and you click it in the wallet selector UI it should show the `WalletNotInstalled` view and not prompt the connection with FiNER.
> 
> I agree, but this is done for compatibility, we can solve this by notifying the user that this is not a Sender and offering to use FiNER, if he refuses, then we will set a flag about this and will not offer FiNER, ok?

Hi @prih, what if some other wallet team wants the users to be directed to their wallet in the same way? 

It doesn't seem fair in this example not to show to the users when they select Sender the`WalletNotInstalled` view first as we have it currently, with the option for them to be redirected to the store to install it.

**Comment by prih** [comment]
> 

this is understandable and logical, but the user has already installed, for example, the Finer extension, for some reason he needed another wallet, that is, for some reason he consciously wants to use both at once? ok, only in this case it ? when our extension is not installed, we do not recommend Finer instead of Sender.

**Comment by andrunovak** [comment]
@AmmarHumackicSQA 

> Hi @prih, the reason behind enforcing the standard is that we ensure everyone has the same conditions for their wallets without negative impact on each other.

Please link this standard so we can find more details there.

**Comment by Olexiy-pt** [comment]
@AmmarHumackicSQA 
> > > If sender is not installed and you click it in the wallet selector UI it should show the `WalletNotInstalled` view and not prompt the connection with FiNER.
> > 
> > 
> > I agree, but this is done for compatibility, we can solve this by notifying the user that this is not a Sender and offering to use FiNER, if he refuses, then we will set a flag about this and will not offer FiNER, ok?
> 
> Hi @prih, what if some other wallet team wants the users to be directed to their wallet in the same way?
> 
> It doesn't seem fair in this example not to show to the users when they select Sender the`WalletNotInstalled` view first as we have it currently, with the option for them to be redirected to the store to install it.


If the user didn't install Sender or Finer, where in the code do you find that we try to block Sender installation?

**Comment by AmmarHumackicSQA** [comment]
> @AmmarHumackicSQA
> 
> > > > If sender is not installed and you click it in the wallet selector UI it should show the `WalletNotInstalled` view and not prompt the connection with FiNER.
> > > 
> > > 
> > > I agree, but this is done for compatibility, we can solve this by notifying the user that this is not a Sender and offering to use FiNER, if he refuses, then we will set a flag about this and will not offer FiNER, ok?
> > 
> > 
> > Hi @prih, what if some other wallet team wants the users to be directed to their wallet in the same way?
> > It doesn't seem fair in this example not to show to the users when they select Sender the`WalletNotInstalled` view first as we have it currently, with the option for them to be redirected to the store to install it.
> 
> If the user didn't install Sender or Finer, where in the code do you find that we try to block Sender installation?

Hi @Olexiy-pt it wasn't mentioned that you block it, but we are saying it should show the user how we currently have it with the`WalletNotInstalled` view. 

To show the option first for the users to be redirected to the store to install Sender if it's not installed already. You can check out our live example: [near.github.io/wallet-selector/](https://near.github.io/wallet-selector/). 

If you have some examples to share of your suggestions we would appreciate it. Thanks!

**Comment by prih** [comment]
> > @AmmarHumackicSQA
> > > > > If sender is not installed and you click it in the wallet selector UI it should show the `WalletNotInstalled` view and not prompt the connection with FiNER.
> > > > 
> > > > 
> > > > I agree, but this is done for compatibility, we can solve this by notifying the user that this is not a Sender and offering to use FiNER, if he refuses, then we will set a flag about this and will not offer FiNER, ok?
> > > 
> > > 
> > > Hi @prih, what if some other wallet team wants the users to be directed to their wallet in the same way?
> > > It doesn't seem fair in this example not to show to the users when they select Sender the`WalletNotInstalled` view first as we have it currently, with the option for them to be redirected to the store to install it.
> > 
> > 
> > If the user didn't install Sender or Finer, where in the code do you find that we try to block Sender installation?
> 
> Hi @Olexiy-pt it wasn't mentioned that you block it, but we are saying it should show the user how we currently have it with the`WalletNotInstalled` view.
> 
> To show the option first for the users to be redirected to the store to install Sender if it's not installed already. You can check out our live example: [near.github.io/wallet-selector/](https://near.github.io/wallet-selector/).
> 
> If you have some examples to share of your suggestions we would appreciate it. Thanks!

well, we will remove the automatic set in window.near, and the user himself can enable it in the settings. @AmmarHumackicSQA - ok?

**Comment by AmmarHumackicSQA** [comment]
> > > @AmmarHumackicSQA
> > > > > > If sender is not installed and you click it in the wallet selector UI it should show the `WalletNotInstalled` view and not prompt the connection with FiNER.
> > > > > 
> > > > > 
> > > > > I agree, but this is done for compatibility, we can solve this by notifying the user that this is not a Sender and offering to use FiNER, if he refuses, then we will set a flag about this and will not offer FiNER, ok?
> > > > 
> > > > 
> > > > Hi @prih, what if some other wallet team wants the users to be directed to their wallet in the same way?
> > > > It doesn't seem fair in this example not to show to the users when they select Sender the`WalletNotInstalled` view first as we have it currently, with the option for them to be redirected to the store to install it.
> > > 
> > > 
> > > If the user didn't install Sender or Finer, where in the code do you find that we try to block Sender installation?
> > 
> > 
> > Hi @Olexiy-pt it wasn't mentioned that you block it, but we are saying it should show the user how we currently have it with the`WalletNotInstalled` view.
> > To show the option first for the users to be redirected to the store to install Sender if it's not installed already. You can check out our live example: [near.github.io/wallet-selector/](https://near.github.io/wallet-selector/).
> > If you have some examples to share of your suggestions we would appreciate it. Thanks!
> 
> well, we will remove the automatic set in window.near, and the user himself can enable it in the settings. @AmmarHumackicSQA - ok?

Hi @prih that would be great. Thanks!

**Comment by prih** [comment]
@AmmarHumackicSQA @DamirSQA we updated extension **v0.1.5**, it doesn't use `window.near` by default

**Comment by DamirSQA** [comment]
Thanks @prih , I'll ask you to merge latest changes with your branch  before we proceed with merging PR.

**Comment by prih** [comment]
> Thanks @prih , I'll ask you to merge latest changes with your branch before we proceed with merging PR.

done

**Review by DamirSQA — APPROVED** [review]

**Comment by prih** [comment]
thx!

**Comment by AmmarHumackicSQA** [comment]
Hi @prih we tried testing the latest version (v0.4.2) of the FiNER wallet and we are getting the "Something went wrong" msg while trying to set up the wallet (tried multiple options: quick start, create a new wallet, and import wallet). Here is the ss of what we are getting:

IMG_0352

Can you take a look at this? Thanks!

**Pull request merged**